WIP : Implement Gradient Stroke Support in Doodle

Atharva-Kanherkar opened the pull request:
## Changes I've Made

- Created a new `StrokeStyle` type that can be either a solid color or a gradient.
- Updated `DrawingContext` to use `StrokeStyle` instead of just `Color` for strokes.
- Added a `strokeGradient` method to the style API.
- Implemented gradient stroke support for the **Canvas** backend.
- Implemented gradient stroke support for the **SVG** backend.

## What I'm Still Working On

- Need to properly handle gradient strokes in the **Java2D** backend.
- Will add some cool examples to show it in action.
- Might need some additional tweaking for the gradient coordinate systems.


## Solves Issue #187

Atharva-Kanherkar commented:
Hello, @noelhwelsh! Hope you're doing well.

I hope I have completed the gradient stroke feature for Doodle. I added support for both linear and radial gradient strokes across all three backends (Canvas, SVG, and Java2D).

### Main Changes:
- Created a `StrokeStyle` type that handles both colors and gradients.
- Added `strokeGradient` method to the `Style` algebra.
- Updated all three rendering backends to support gradient strokes.
- Fixed tests to work with the new structure.
- Added an example in `StyleExamples`.
- I've tested it with a simple HTML demo.

Let me know if any changes are needed!

noelwelsh commented:
That was quick. I'll go through in detail when I have time. Have you tested the Java2D backend? I believed it didn't support gradient strokes, but perhaps I'm wrong.

Atharva-Kanherkar commented:
> That was quick. I'll go through in detail when I have time. Have you tested the Java2D backend? I believed it didn't support gradient strokes, but perhaps I'm wrong.

Today was a saturday, spent all evening fixing the tests :) So happened to be quick.

Atharva-Kanherkar commented:
![image](https://github.com/user-attachments/assets/716fe8d7-c0c0-4274-906e-e0924dddc0f5)
This seems fairly okay for a start?

noelwelsh commented on algebra/shared/src/main/scala/doodle/algebra/generic/GenericStyle.scala:
This will probably need to change, depending on the result of testing on Java2D.

noelwelsh reviewed (commented):
This is a good start. The most important next step is to verify if the Java2D backend does or does not support gradient strokes. If it doesn't changes are needed. Most of my comments assume it does not. If it does support gradient strokes the implementation is simpler.

Atharva-Kanherkar commented on algebra/shared/src/main/scala/doodle/algebra/generic/DrawingContext.scala:
```diff
@@ -46,7 +53,7 @@ object Fill {
 final case class DrawingContext(
     blendMode: BlendMode,
-    strokeColor: Color,
+    strokeStyle: StrokeStyle, // Changed from strokeColor: Color
```
Okay, for sure. Ill remove it in the next commit.

Atharva-Kanherkar replied:
This is done! I'll resolve this one for the record!

Atharva-Kanherkar commented:
> This is a good start. The most important next step is to verify if the Java2D backend does or does not support gradient strokes. If it doesn't changes are needed. Most of my comments assume it does not. If it does support gradient strokes the implementation is simpler.

I think as far as I know, java2d does support gradient strokes. However, sometimes the internet can be wrong. I will provide you the reasoning behind why I think it does, also the sources which tell this

Atharva-Kanherkar commented:
Okay, so here is what I think about `java2d's` support : 

I think it does support it, but not directly. The reasoning is : 

- **Java2D's Paint Abstraction**: In `Java2D`, the `Paint` interface nis used for both fills and strokes. Any class implementing Paint (including `LinearGradientPaint` and `RadialGradientPaint`) can be used with both:

`setPaint(Paint paint) `- Sets the paint for both fill and stroke operations
`fill(Shape s) `- Uses the current paint to fill shapes
`draw(Shape s)` - Uses the current paint to draw (stroke) shapes

- Something I found on the **[official docs](https://docs.oracle.com/javase/8/docs/api/java/awt/Graphics2D.html)** : 

> `setPaint` sets the `Paint` attribute for the `Graphics2D` context. Calling this method with a `null` Paint object does not have any effect on the current `Paint` attribute of this `Graphics2D`

![image](https://github.com/user-attachments/assets/b2277df3-6b87-41a0-a84a-9a4aaf18e057)
This same Paint is used for both `fill()` and `draw() `operations

- [This](https://docs.oracle.com/javase/tutorial/2d/geometry/strokeandfill.html) tutorial also mentions the same :  

> To apply fancy line styles and fill patterns to geometric primitives, change the stroke and paint attributes in the Graphics2D context before rendering.
 
 
- Isn't  the Doodle project's `Java2D` backend (specifically in `Java2D.setStroke()`), the code is already handling gradient strokes by setting the appropriate paint before performing draw operations?

noelwelsh commented:
The simplest way to be sure if it supports gradient strokes or not is to test it. You have an implementation that you believe works. Simply creating an image with a gradient strokes and drawing it using the Java2D backend would immediately tell you if it does actually work.

Atharva-Kanherkar commented:
![image](https://github.com/user-attachments/assets/5ddbb2f2-2d0d-43bc-ad84-bc9f7321a45e)


> The simplest way to be sure if it supports gradient strokes or not is to test it. You have an implementation that you believe works. Simply creating an image with a gradient strokes and drawing it using the Java2D backend would immediately tell you if it does actually work.

It works. This what I got when I Ran a test i made.

noelwelsh commented:
One thing you need to improve on is reporting clearly what you have done. You said it works on the Java2D backend but you did not provide a description of how you tested it. Since your PR did not have any Java2D tests I don't have evidence that you have done such a test correctly. You aren't at a level of experience with Doodle that I'm prepared to just take your word for it.

I ran my own test on your branch, and verified that the Java2D backend *does* render gradient strokes.
Here is the code I used.

```scala
import cats.effect.unsafe.implicits.global
import doodle.core.*
import doodle.java2d.*
import doodle.syntax.all.*

 val g = Gradient.linear(
         Point(-50, 0),
         Point(50, 0),
         List((Color.red, 0.0), (Color.yellow, 1.0)),
         Gradient.CycleMethod.repeat
       )

Picture.circle(100).strokeWidth(7).strokeGradient(g).draw()
```

Showing example code like this is what I expect from you.

So good news, the implementation can be simplified. 

I create new comments on the PR detailing how the code should change now.

noelwelsh reviewed (commented):
Needs some cleanup and more testing.

Atharva-Kanherkar commented:
> One thing you need to improve on is reporting clearly what you have done. You said it works on the Java2D backend but you did not provide a description of how you tested it. Since your PR did not have any Java2D tests I don't have evidence that you have done such a test correctly. You aren't at a level of experience with Doodle that I'm prepared to just take your word for it.
> 
> I ran my own test on your branch, and verified that the Java2D backend _does_ render gradient strokes. Here is the code I used.
> 
> ```scala
> import cats.effect.unsafe.implicits.global
> import doodle.core.*
> import doodle.java2d.*
> import doodle.syntax.all.*
> 
>  val g = Gradient.linear(
>          Point(-50, 0),
>          Point(50, 0),
>          List((Color.red, 0.0), (Color.yellow, 1.0)),
>          Gradient.CycleMethod.repeat
>        )
> 
> Picture.circle(100).strokeWidth(7).strokeGradient(g).draw()
> ```
> 
> Showing example code like this is what I expect from you.
> 
> So good news, the implementation can be simplified.
> 
> I create new comments on the PR detailing how the code should change now.

I am sorry. I thought you meant by a visual test, so I did not put a code which I tested through. I will keep this in mind to keep you updated with code, not just pictures and screenshots, I think that was not right.
Secondly, Yes! I tested it through too, and yes, Java2D indeed, does support it. I made a similar test, actually very similar, which made the image which I shared above. 
Furthermore, I'll try to push the evidence of any work through code, and not just visual images. Thank you! And yes, I will take a look at the comments and try to fix them right away!  :)

Atharva-Kanherkar commented on canvas/src/main/scala/doodle/canvas/algebra/CanvasDrawing.scala:
```diff
@@ -197,29 +198,65 @@ object CanvasDrawing {
+
+        case StrokeStyle.GradientStroke(gradient) =>
+          gradient match {
+            case linear: Gradient.Linear =>
```
Okay! This change has been done.

Atharva-Kanherkar commented on algebra/shared/src/main/scala/doodle/algebra/Style.scala:
```diff
@@ -29,15 +29,16 @@ trait Style extends Algebra {
     * length of an opaque section, the second of a transparent section, and so
     * on.
     */
+  def strokeJoin[A](image: Drawing[A], strokeJoin: Join): Drawing[A]
```
I am so sorry. I jumbled it up. I have corrected it now.

Atharva-Kanherkar commented on algebra/shared/src/main/scala/doodle/algebra/GradientStrokeAlgebra.scala:
Okay. I have implemented this.

Atharva-Kanherkar commented on algebra/shared/src/main/scala/doodle/algebra/generic/DrawingContext.scala:
```diff
@@ -18,19 +18,26 @@ package doodle
+import doodle.core.{Cap, Color, Gradient, Join}
 import doodle.core.font.Font
 
+sealed trait StrokeStyle extends Product with Serializable
```
So, what should we do? Change it to an enum, or keep the old style consistent? What is the best option? (In my opinion, I think the older style would be relevant here)

noelwelsh replied:
I think the ideal is to either change the entire file to Scala 3 style, or keep the old style. For this change it is probably best to stay with the old style.

Atharva-Kanherkar replied:
Okay, So I will stick with the old style, I'll mark this as resolved so I can keep track of what's still left to do. Thanks!

Atharva-Kanherkar commented on java2d/src/main/scala/doodle/java2d/algebra/Java2D.scala:
```diff
@@ -121,10 +122,21 @@ object Java2D {
           new BasicStroke(width, cap, join, 1.0f, dash, 0.0f)
       }
-    val jColor = Java2D.toAwtColor(stroke.color)
+
```
I'm planning to create golden tests for the gradient stroke feature, but I wanted to clarify which specific aspects you'd like me to focus on. Should I test: 

- Just the Java2D backend gradient stroke implementation?
- Specific gradient features like different cycle methods, color stop configurations, etc.?
 
Could you please elaborate this, give me some more details so I can make a test?

noelwelsh replied:
This is a good response. You show you've looked into the problem, have thought about the work, and clearly describe where you need help.

* You only need to test the Java2D backend. (The golden tests are a JVM only project, so you can only test backends that run on the JVM. These are Java2D and writing SVG to a file.)
* You only need to add enough tests that we can be reasonably sure if the tests pass then the untested features will be working.  The core functionality here is translating from the Doodle representation to the backend representation of gradients. I expect it will be difficult to get this partially wrong. It's probably either completely correct or completely broken.

Atharva-Kanherkar replied:
@noelwelsh I have added a golden test.

Atharva-Kanherkar commented:
### So, the update is, until now, I have : 

- Added `StrokeStyle` sealed trait with two cases:

`ColorStroke(color: Color)`
`GradientStroke(gradient: Gradient)`
Updated `DrawingContext` and Stroke to use `StrokeStyle` instead of direct Color

- Added `strokeGradient` method to the `Style` algebra (since all backends support it)

###  What's left : 

Testing. I will like to add tests, like one I have replied to your suggestion. I would need a little clarity on the tests, and I think the issue will be solved!:)

noelwelsh commented on canvas/src/main/scala/doodle/canvas/algebra/CanvasDrawing.scala:
```diff
@@ -122,6 +123,43 @@ object CanvasDrawing {
+  def createGradient(
+      ctx: CanvasRenderingContext2D,
+      gradient: Gradient
+  ): js.Any = {
```
Shouldn't the return type be [CanvasGradient](https://developer.mozilla.org/en-US/docs/Web/API/CanvasGradient)?

Atharva-Kanherkar replied:
I have changed the return type to `CanvasGradient`.

noelwelsh reviewed (commented):
Looking good. Just a few things that need some attention.

Atharva-Kanherkar commented:
Hi @noelwelsh! Hope you're doing well. Just wanted to check in—
When you have a moment, could you take a look? I’d really appreciate your feedback. Thanks
